File: node/src/http-client.ts
export type UserRole = 'merchant' | 'carrier';

export interface RequestInterceptor {
    onRequest?: (url: string, options: RequestInit) => RequestInit | Promise<RequestInit>;
    onResponse?: (response: Response) => Response | Promise<Response>;
    onError?: (error: Error) => void;
}

export interface ZymeupConfig {
    baseUrl: string;
    token?: string;
    timeout: number;
    role?: UserRole;
    carrierCode?: string;
    maxRetries?: number;
    retryDelayMs?: number;
    interceptors?: RequestInterceptor;
}

// Backward compatibility alias
export type ShipzyConfig = ZymeupConfig;

export const DEFAULT_CONFIG: Partial<ZymeupConfig> = {
    baseUrl: 'https://api.zymeup.com',
    timeout: 30000,
    role: 'merchant',
    maxRetries: 3,
    retryDelayMs: 1000,
};

export class ZymeupError extends Error {
    constructor(message: string, public statusCode: number) {
        super(message);
        this.name = 'ZymeupError';
    }
}

export class ZymeupAuthError extends ZymeupError {
    constructor(message: string) {
        super(message, 401);
        this.name = 'ZymeupAuthError';
    }
}

export class HttpClient {
    protected config: ZymeupConfig;

    constructor(config: Partial<ZymeupConfig> = {}) {
        this.config = { ...DEFAULT_CONFIG, ...config } as ZymeupConfig;
    }

    setToken(token: string): void {
        this.config.token = token;
    }

    setConfig(config: Partial<ZymeupConfig>): void {
        if (config.baseUrl) {
            this.config.baseUrl = config.baseUrl;
        }
        if (config.token) {
            this.config.token = config.token;
        }
        if (config.timeout) {
            this.config.timeout = config.timeout;
        }
        if (config.role) {
            this.config.role = config.role;
        }
        if (config.carrierCode) {
            this.config.carrierCode = config.carrierCode;
        }
        if (config.maxRetries !== undefined) {
            this.config.maxRetries = config.maxRetries;
        }
        if (config.retryDelayMs !== undefined) {
            this.config.retryDelayMs = config.retryDelayMs;
        }
        if (config.interceptors !== undefined) {
            this.config.interceptors = config.interceptors;
        }
    }

    protected getAuthHeader(): string {
        if (this.config.role === 'carrier' && this.config.carrierCode && this.config.token) {
            return `Bearer ${this.config.carrierCode}:${this.config.token}`;
        }
        return `Bearer ${this.config.token || ''}`;
    }

    protected async request<T>(path: string, method: string = 'GET', body?: unknown, isFormData?: boolean): Promise<T> {
        const maxRetries = this.config.maxRetries ?? 3;
        const retryDelayMs = this.config.retryDelayMs ?? 1000;
        const interceptor = this.config.interceptors;
        const url = `${this.config.baseUrl.replace(/\/$/, '')}${path}`;

        let lastError: Error | undefined;

        for (let attempt = 0; attempt <= maxRetries; attempt++) {
            const headers: Record<string, string> = {};
            if (!isFormData) {
                headers['Content-Type'] = 'application/json';
            }
            if (this.config.token) {
                headers['Authorization'] = this.getAuthHeader();
            }

            let options: RequestInit = {
                method,
                headers,
                body: isFormData ? (body as FormData) : (body ? JSON.stringify(body) : undefined),
            };

            if (interceptor?.onRequest) {
                options = await interceptor.onRequest(url, options);
            }

            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

            try {
                let response = await fetch(url, {
                    ...options,
                    signal: controller.signal,
                });

                if (interceptor?.onResponse) {
                    response = await interceptor.onResponse(response);
                }

                if (response.status === 401) throw new ZymeupAuthError('Unauthorized');

                if (response.status >= 500 && attempt < maxRetries) {
                    await this.sleep(retryDelayMs * 2 ** attempt);
                    continue;
                }

                if (!response.ok) {
                    const errorData = await response.json().catch(() => ({}));
                    throw new ZymeupError(errorData.message || `HTTP ${response.status}`, response.status);
                }

                return response.json();
            } catch (error) {
                lastError = error instanceof Error ? error : new Error(String(error));

                const isNetworkError = lastError.name === 'AbortError' || lastError.message.includes('fetch');
                const isServerError = lastError instanceof ZymeupError && lastError.statusCode >= 500;

                if ((isNetworkError || isServerError) && attempt < maxRetries) {
                    await this.sleep(retryDelayMs * 2 ** attempt);
                    continue;
                }

                if (interceptor?.onError) {
                    interceptor.onError(lastError);
                }

                throw lastError;
            } finally {
                clearTimeout(timeoutId);
            }
        }

        if (interceptor?.onError && lastError) {
            interceptor.onError(lastError);
        }

        throw lastError;
    }

    private sleep(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    protected buildQuery(params: Record<string, unknown>): string {
        const parts: string[] = [];
        for (const [key, value] of Object.entries(params)) {
            if (value !== undefined && value !== null) {
                parts.push(`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`);
            }
        }
        return parts.length ? '?' + parts.join('&') : '';
    }
}

File: rust/src/http_client.rs
use std::fmt;

use reqwest::{Client, Method};
use serde::de::DeserializeOwned;
use tokio::time::Duration;

use crate::error::{Result, ShipzyError};

#[derive(Debug, thiserror::Error)]
pub enum HttpClientError {
    #[error("HTTP error {status}: {message}")]
    Http { status: u16, message: String },
    #[error("Request error: {0}")]
    Reqwest(#[from] reqwest::Error),
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),
}

impl From<HttpClientError> for ShipzyError {
    fn from(err: HttpClientError) -> Self {
        match err {
            HttpClientError::Http { status, message } => ShipzyError::Http { status, message },
            HttpClientError::Reqwest(e) => ShipzyError::Reqwest(e),
            HttpClientError::Json(e) => ShipzyError::Json(e),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ZymeupConfig {
    pub base_url: String,
    pub token: Option<String>,
    pub timeout_seconds: u64,
    pub role: UserRole,
    pub carrier_code: Option<String>,
}

impl Default for ZymeupConfig {
    fn default() -> Self {
        Self {
            base_url: "https://api.zymeup.com".to_string(),
            token: None,
            timeout_seconds: 30,
            role: UserRole::Merchant,
            carrier_code: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum UserRole {
    Merchant,
    Carrier,
}

impl Default for UserRole {
    fn default() -> Self {
        Self::Merchant
    }
}

#[derive(Debug, Clone)]
pub struct HttpClient {
    client: Client,
    config: ZymeupConfig,
}

impl HttpClient {
    pub fn new(config: ZymeupConfig) -> Result<Self> {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.timeout_seconds))
            .build()
            .map_err(|e| ShipzyError::Reqwest(e))?;
        Ok(Self { client, config })
    }

    pub fn set_token(&mut self, token: &str) {
        self.config.token = Some(token.to_string());
    }

    pub fn config(&self) -> &ZymeupConfig {
        &self.config
    }

    pub fn base_url(&self) -> &str {
        &self.config.base_url
    }

    pub fn token(&self) -> &Option<String> {
        &self.config.token
    }

    pub fn client(&self) -> &Client {
        &self.client
    }

    pub async fn request<T: DeserializeOwned>(
        &self,
        path: &str,
        method: Method,
        body: Option<serde_json::Value>,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url.trim_end_matches('/'), path);
        let mut req = self.client.request(method, &url);

        if let Some(ref token) = self.config.token {
            let auth = if self.config.role == UserRole::Carrier
                && self.config.carrier_code.is_some()
            {
                let carrier_code = self.config.carrier_code.as_ref().unwrap();
                format!("Bearer {}:{}", carrier_code, token)
            } else {
                format!("Bearer {}", token)
            };
            req = req.header(reqwest::header::AUTHORIZATION, auth);
        }

        req = req.header(reqwest::header::CONTENT_TYPE, "application/json");

        if let Some(body) = body {
            req = req.json(&body);
        }

        let resp = req.send().await?;
        let status = resp.status();

        if status == 401 {
            return Err(ShipzyError::Auth);
        }

        if !status.is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(ShipzyError::Http {
                status: status.as_u16(),
                message: text,
            });
        }

        let t: T = resp.json().await?;
        Ok(t)
    }
}

impl fmt::Display for UserRole {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            UserRole::Merchant => write!(f, "merchant"),
            UserRole::Carrier => write!(f, "carrier"),
        }
    }
}
